src/binder/bind/bind_reading_clause.cpp

```diff
@@ -137,86 +137,136 @@ std::unique_ptr<BoundReadingClause> Binder::bindLoadFrom(
+    return boundLoadFrom;
+}
+
+expression_vector Binder::createColumnExpressions(common::ReaderConfig& readerConfig,
+    const std::vector<std::string>& columnNames,
+    const std::vector<std::unique_ptr<common::LogicalType>>& columnTypes) {
+    expression_vector columns;
+    for (auto i = 0u; i < columnTypes.size(); ++i) {
+        auto columnName = columnNames[i];
+        auto columnType = columnTypes[i].get();
+        readerConfig.columnNames.push_back(columnName);
+        readerConfig.columnTypes.push_back(columnType->copy());
+        columns.push_back(createVariable(columnName, *columnType));
+    }
+    return columns;
+}
+
+void Binder::validateColumnTypes(const std::vector<std::string>& columnNames,
+    const std::vector<std::unique_ptr<LogicalType>>& expectedColumnTypes,
+    const std::vector<std::unique_ptr<LogicalType>>& detectedColumnTypes) {
+    assert(expectedColumnTypes.size() == detectedColumnTypes.size());
+    for (auto i = 0; i < expectedColumnTypes.size(); ++i) {
+        if (*expectedColumnTypes[i] != *detectedColumnTypes[i]) {
+            throw BinderException(
+                StringUtils::string_format("Column {} type mismatch. Expect {} but get {}.",
+                    columnNames[i], LogicalTypeUtils::dataTypeToString(*expectedColumnTypes[i]),
+                    LogicalTypeUtils::dataTypeToString(*detectedColumnTypes[i])));
+        }
+    }
+}
+
+void Binder::validateNumColumns(uint32_t expectedNumber, uint32_t detectedNumber) {
+    if (detectedNumber == 0) {
+        return; // Empty CSV. Continue processing.
+    }
+    if (expectedNumber != detectedNumber) {
+        throw BinderException(
+            StringUtils::string_format("Number of columns mismatch. Detect {} but expect {}.",
+                detectedNumber, expectedNumber));
+    }
+}
+
+void Binder::sniffFiles(const common::ReaderConfig& readerConfig,
+    std::vector<std::string>& columnNames,
+    std::vector<std::unique_ptr<common::LogicalType>>& columnTypes) {
+    assert(readerConfig.getNumFiles() > 0);
+    sniffFile(readerConfig, 0, columnNames, columnTypes);
+    for (auto i = 1; i < readerConfig.getNumFiles(); ++i) {
+        std::vector<std::string> tmpColumnNames;
+        std::vector<std::unique_ptr<LogicalType>> tmpColumnTypes;
+        sniffFile(readerConfig, i, tmpColumnNames, tmpColumnTypes);
+        switch (readerConfig.fileType) {
+        case FileType::CSV: {
+            validateNumColumns(columnTypes.size(), tmpColumnTypes.size());
+        }
+        case FileType::PARQUET: {
+            validateNumColumns(columnTypes.size(), tmpColumnTypes.size());
+            validateColumnTypes(columnNames, columnTypes, tmpColumnTypes);
+        } break;
+        case FileType::NPY: {
+            validateNumColumns(1, tmpColumnTypes.size());
+            columnNames.push_back(tmpColumnNames[0]);
+            columnTypes.push_back(tmpColumnTypes[0]->copy());
+        } break;
+        default:
+            break;
+        }
```

Comment:
```suggestion
        case FileType::TURTLE:
            break
        default:
            // LCOV_EXCL_START
            throw NotImplementedException("Binder::sniffFiles");
            // LCOV_EXCL_END
        }
```

This will make the behaviour less confusing to the next person to add a file type.